src/ccs/ir/circuit.rs

```diff
@@ -0,0 +1,579 @@
+use crate::{
+    ccs::compiler::step_selector::{SelectorsForALLMatrix, SelectorsForALLSteps, StepSelector},
+    field::Field,
+    util::UUID,
+};
+
+pub type MatrixCoeffsAndOffset<F> = Vec<Vec<(CoeffsForProds<F>, usize)>>;
+pub type SelectorsOffsetAndCoeff<F> = Vec<Vec<(usize, F)>>;
+
+// ((step_idx, step_UUID), assignment_offset)
+pub type AssignmentOffsets = HashMap<(usize, u128), usize>;
+
+pub fn hadamard_product<F: Field + From<u64> + Hash>(vec1: &[F], vec2: &[F]) -> Vec<F> {
+    assert_eq!(vec1.len(), vec2.len());
+    vec1.iter()
+        .zip(vec2.iter())
+        .map(|(&v1, &v2)| v1 * v2)
+        .collect()
+}
+
+pub fn vec_add<F: Field + From<u64> + Hash>(vec: &[Vec<F>]) -> Vec<F> {
+    assert!(vec.len() > 1);
+    vec.iter().fold(vec![F::ZERO; vec[0].len()], |acc, vec| {
+        acc.iter().zip(vec.iter()).map(|(&a, &v)| a + v).collect()
+    })
+}
+
+// input vector z = [1, x, w]
+pub struct Z<F: Field + From<u64>> {
+    pub n: usize,
+    pub l: usize,
+    pub assignments: Vec<F>,
+    pub public_inputs: Vec<F>,
+}
+
+impl<F: Field + From<u64> + Hash> Z<F> {
+    pub fn new(n: usize, l: usize) -> Self {
+        assert!(n > l);
+        Self {
+            n,
+            l,
+            assignments: Default::default(),
+            public_inputs: Default::default(),
+        }
+    }
+
+    pub fn write(
+        &mut self,
+        inputs: &HashMap<(usize, UUID), F>,
+        witnesses: &[(UUID, HashMap<UUID, F>)],
+        assign_pos: &AssignmentOffsets,
+    ) {
+        assert_eq!(inputs.len(), self.l);
+        assert_eq!(assign_pos.len(), self.n);
+        let witness_len = self.n - self.l - 1;
+        let mut witness_values = vec![F::ZERO; witness_len];
+        let mut public_values = vec![F::ZERO; self.l];
+        for ((step_idx, signal_id), idx) in assign_pos.iter() {
+            if *signal_id == 0 {
+                continue;
+            }
+            if *idx < self.n - self.l {
+                witness_values[*idx - 1] = *witnesses[*step_idx].1.get(signal_id).unwrap();
+            } else {
+                public_values[*idx - witness_len - 1] =
+                    *inputs.get(&(*step_idx, *signal_id)).unwrap();
+            }
+        }
+        self.assignments = witness_values;
+        self.public_inputs = public_values;
+    }
+
+    pub fn write_with_values(&mut self, inputs: &[F], witnesses: &[F]) {
+        assert_eq!(inputs.len(), self.l);
+        assert_eq!(witnesses.len(), self.n - self.l - 1);
+
+        self.public_inputs.clone_from(&inputs.to_owned());
+        self.assignments.clone_from(&witnesses.to_owned());
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Matrix<F> {
+    n: usize,
+    m: usize,
+    values: Vec<Vec<F>>,
+}
+
+impl<F: Field> Matrix<F> {
+    pub fn new(n: usize, m: usize) -> Self {
+        Self {
+            n,
+            m,
+            values: vec![vec![F::ZERO; n]; m],
+        }
+    }
+
+    pub fn write(&mut self, values: &[(usize, usize, F)]) {
```

Comment:
Again `write` is only necessary if Matrix is modified.

Reply:
ok